classes/map.js

```diff
@@ -92,6 +97,37 @@ Map.prototype.dragPan = {
 
 Map.prototype.project = function() {}
 
-Map.prototype.queryRenderedFeatures = function() {
-  return [];
+Map.prototype.queryRenderedFeatures = function(bbox, queryParams) {
+  var tbb = [];
+  if (bbox[0].x !== undefined) {
+    tbb = [
+      Math.min(bbox[0].x, bbox[1].x),
+      Math.min(bbox[0].y, bbox[1].y),
+      Math.max(bbox[0].x, bbox[1].y),
+      Math.max(bbox[0].x, bbox[1].y)
+    ]
+  } else {
+    tbb = [
+      Math.min(bbox[0][0], bbox[1][0]),
+      Math.min(bbox[0][1], bbox[1][1]),
+      Math.max(bbox[0][0], bbox[1][0]),
+      Math.max(bbox[0][1], bbox[1][1])
+    ];
+  }
+
+  var bpoly = bboxPolygon(tbb);
+  var features = Object.keys(this._sources).reduce((memo, name) => memo.concat(this._sources[name].data.features), []);
+  features = features.filter(feature => {
+      if (feature.geometry.type === 'Point') {
+        feature = buffer(feature, .00000001, 'kilometers');
+      }
+      var fpoly = envelope({
+        type: 'FeatureCollection',
+        features: [feature]
+      });
+      var merged = turfUnion(fpoly, bpoly);
+      return merged.geometry.type === 'Polygon';
```

Comment:
I do not understand what is happening with the envelope, union, and filtering out non-Polygons. Could you explain?